### PR TITLE
TR-1: Pull images from dockerhub (exo-setup)

### LINKED ISSUE #66: Determine if service is currently in development (exo-setup)
In `exo-setup`, we currently use the `location` field of `application.yml` as a directory to build a docker image from. To account for cases where a user may be using some pre-made services from the future exosphere "service-store", we should only build the image if the `location` field is there (indicating that the user is actively developing that service), and if not, but there is a `docker_image` field, we should pull the image at that time.

### PR DESCRIPTION
**Pull request by trushton**
resolves #66

If a service in `application.yml` has a `docker-image` field, but no `location`, the image should be pulled from docker hub, but not set up in the way that we do for local services.

@kevgo @hugobho

### CODE REVIEW

File: exo-setup/features/setup.feature
Comment (kevgo) [inline exo-setup/features/setup.feature]: Scenario: setting up an application with external Docker images

File: exo-setup/features/setup.feature
Comment (kevgo) [inline exo-setup/features/setup.feature]: Not sure if we need to re-test this here again. I think this scenario is only for verifying that it downloads the Docker images. Let's remove this `Then it has created the folders` step here.

File: exo-setup/src/docker-setup.ls
Comment (hugoesthere) [inline exo-setup/src/docker-setup.ls]: I think this has an extra tab to it

**Comment by hugobho** [comment]
What happens when both the `location` and `docker-image` fields are listed, does it use only the local version of the service?

File: exosphere-shared/example-apps/docker-service/application.yml
Comment (hugoesthere) [inline exosphere-shared/example-apps/docker-service/application.yml]: What happens if both the `docker_image` and `location` field are defined? Does it use only the local version of the service? If so shouldn't we only define one or the other?

Comment (trushton) [reply exosphere-shared/example-apps/docker-service/application.yml]: Thats right, it uses the local version. Reason being that if you have the repository locally, it is reasonable to believe that you are developing it and want to use your latest changes.

Comment (hugoesthere) [reply exosphere-shared/example-apps/docker-service/application.yml]: What's the point of defining `docker_image` as well then? Isn't the image name generated in `exo-setup`? Having that in two places might be confusing. It seems like it's not being used here anyways?

Comment (trushton) [reply exosphere-shared/example-apps/docker-service/application.yml]: Its not being used here, but it would still be beneficial to have the field here for when the time comes to deploy a service to dockerhub, we wouldn't need to generate the name again, it could just be grabbed from here. 

Overall I thought it best to allow for any configuration of having the fields, you could have both and it use the location, if you just have one it uses that method, and if you have neither, an error is signaled.

Does that seem right? or would it be better to have a user remove the `docker_image` if they want to build it themselves?

Comment (hugoesthere) [reply exosphere-shared/example-apps/docker-service/application.yml]: I don't think this field can be used to deploy a service to Dockerhub right? Because it's `exo-setup` that generates the Docker images, and that uses a certain naming convention. The user can name the image whatever he/she wants in this `docker_image` field, but that won't be the name of the image that was built in `exo setup` (or the one that will be deployed in `exo deploy`).

I think it makes sense to allow for any configuration, and we can do that by having one of  either `docker_image` or `location`. That would still allow the user to use either a remote docker image or their local service: you list the `location` field, you use a local service; you list the `docker_image` field, you use a remote service. But when both `docker_image` and `location` are included, the `docker_image` field will always be ignored it seems like. Does that make sense? What do you think?

Or maybe it's a bit confusing that we call one of the fields `location` when a service's location can be either local or remote..

File: exo-setup/src/docker-setup.ls
Comment (kevgo) [inline exo-setup/src/docker-setup.ls]: Extract this into a `start-external-service` method to make this more readable.

File: exosphere-shared/example-apps/docker-service/application.yml
Comment (kevgo) [inline exosphere-shared/example-apps/docker-service/application.yml]: Let's call this app `app-with-external-docker-images`. Also change the directory name.

File: exosphere-shared/example-apps/docker-service/application.yml
Comment (kevgo) [inline exosphere-shared/example-apps/docker-service/application.yml]: There should be an `external-service` entry here, which uses the external Docker image.

File: exosphere-shared/example-apps/docker-service/application.yml
Comment (kevgo) [reply exosphere-shared/example-apps/docker-service/application.yml]: Let's follow the principle of least surprise here, and do the simple and straightforward thing that causes the least amount of confusion. I think that is to have only one or the other setting, but not both. Later we can add extra logic, documentation, and testing to allow for both settings being present at the same time.

File: exosphere-shared/example-apps/docker-service/local-service/features/create-many.feature
Comment (kevgo) [inline exosphere-shared/example-apps/docker-service/local-service/features/create-many.feature]: Please slim this service down to the simplest possible service. Only one functionality, and only one test for it. It could just be a ping service, which replies "pong" when receiving "ping".

**Comment by trushton** [comment]
@kevgo This should be good now

File: exo-setup/src/app-setup.ls
Comment (kevgo) [inline exo-setup/src/app-setup.ls]: Move this line to line 28, so that you first deal with setups, and then down here with docker-setups.

File: exo-setup/src/docker-setup.ls
Comment (kevgo) [inline exo-setup/src/docker-setup.ls]: Too many spaces before `=>`

File: exosphere-shared/example-apps/app-with-external-docker-images/local-service/server.ls
Comment (kevgo) [inline exosphere-shared/example-apps/app-with-external-docker-images/local-service/server.ls]: No need for the comment here

File: exo-setup/features/setup.feature
Comment (kevgo) [inline exo-setup/features/setup.feature]: What does "aquired Docker images" mean here? Does it download a Docker image for the `local-service`?

Comment (kevgo) [reply exo-setup/features/setup.feature]: I'm not sure we need the local-service here at all. Isn't this scenario supposed to specify how applications with external Docker images work? I think an app with a single service that uses an external Docker image is enough here.

**Comment by trushton** [comment]
@kevgo

**Review by kevgo — COMMENTED** [review]
Some preliminary comments from last week. This PR is huge now, probably contains a ton of unrelated changes. Can you please sync the branch?

**Comment by trushton** [comment]
I did a `git sync`, I think the issue is that there were sub-branches off of this one that got approved and then merged in, so this one looks as though it has a lot of changes. 
Either way, the comments have been addressed and this should be ready for another look @kevgo

**Comment by kevgo** [comment]
This PR contains elements from #136, which has been merged into `tr-1` instead of master. I think the branch `tr-organize-steps` was a child of this branch. You should have shipped this branch first, then `tr-organize-steps`.

**Comment by kevgo** [comment]
The solution is:
* create a new branch containing the changes from `tr-organize-steps` 
* check out `tr-1` and run `git revert 027304657094aeffe7d5c4407d7c477b34d17cf8`. This should undo the merge of `tr-organize-steps` into this branch here.
* we review this branch and ship it
* you ship the new `tr-organize-steps` branch.

**Comment by trushton** [comment]
@kevgo

File: exo-setup/src/docker-setup.ls
Comment (kevgo) [inline exo-setup/src/docker-setup.ls]: unused module?

File: exo-setup/src/docker-setup.ls
Comment (kevgo) [inline exo-setup/src/docker-setup.ls]: You should check the exit code here, and display at least an error message when something went wrong. Or bubble the error up and handle it higher up in the call chain via `..on 'ended', done`.

File: exo-setup/src/docker-setup.ls
Comment (kevgo) [inline exo-setup/src/docker-setup.ls]: unnecessary return

File: exo-setup/src/docker-setup.ls
Comment (kevgo) [inline exo-setup/src/docker-setup.ls]: Handle the error case first via a guard clause.

File: exosphere-shared/example-apps/app-with-external-docker-images/application.yml
Comment (kevgo) [inline exosphere-shared/example-apps/app-with-external-docker-images/application.yml]: An application whose services are Dockerhub images

File: exosphere-shared/example-apps/app-with-external-docker-images/application.yml
Comment (kevgo) [inline exosphere-shared/example-apps/app-with-external-docker-images/application.yml]: 0.0.0

**Comment by trushton** [comment]
@kevgo Latest comments have been addressed

File: exo-setup/src/service-setup.ls
Comment (kevgo) [inline exo-setup/src/service-setup.ls]: Please fix this merge conflict

**Review by kevgo — APPROVED** [review]
Good to go once the merge conflict is resolved

**Pull request closed without merging**
